libs/portals/admin/ids-admin/src/components/Client/Permissions.tsx

```diff
@@ -1,86 +1,110 @@
-    id: '@island.is/auth/admin:full',
-    description:
-      'Full access to authorization admin something description here',
-    api: 'Island.is APIs',
-    locked: false,
-  },
-  {
-    label: 'Skattskýrslur',
-    id: '@skatturinn.is/skattskyrslur',
-    description:
-      'Full access to authorization admin something description here',
-    api: 'Skatturinn',
-    locked: true,
-  },
-  {
-    label: 'Staða og hreyfingar',
-    id: '@island.is/finance',
-    description:
-      'Skoða stöðu við ríkissjóð og stofnanir, hreyfingar, greiðsluseðla og greiðslukvittanir.',
-    api: 'Island.is APIs',
-    locked: false,
-  },
-  {
-    label: 'Full Access',
-    id: '@island.is/auth/admin',
-    description:
-      'Full access to authorization admin something description here',
-    api: 'Island.is APIs',
-    locked: false,
-  },
-]
-
-function Permissions({ data = mockData }: PermissionsProps) {
-  const { formatMessage } = useLocale()
+function Permissions({ allowedScopes }: PermissionsProps) {
+  const { formatMessage, locale } = useLocale()
   const [isModalVisible, setIsModalVisible] = useState(false)
+  const params = useParams()
+  const [permissions, setPermissions] = useState<AuthAdminClientAllowedScope[]>(
+    allowedScopes ?? [],
+  )
+  const actionData = useActionData() as EditApplicationResult<
+    typeof schema[typeof ClientFormTypes.permissions]
+  >
+
+  const actionDataRef = useRef(actionData?.data)
+
+  useEffect(() => {
+    if (actionData?.intent === ClientFormTypes.permissions) {
+      if (!isEqual(actionData?.data, actionDataRef?.current)) {
+        if (actionData?.data) {
+          setAddedScopes([])
+          setRemovedScopes([])
+        }
+        if (actionData?.globalError) {
+          toast.error(formatMessage(m.globalErrorMessage))
+        }
+      }
+    }
+  }, [actionData])
+
+  const {
+    addedScopes,
+    removedScopes,
+    setRemovedScopes,
+    setAddedScopes,
+  } = useContext(ClientContext)
+
+  const tenant = params['tenant']
+
+  const handleModalOpen = () => {
+    setIsModalVisible(true)
+  }
 
   const handleModalClose = () => {
     setIsModalVisible(false)
   }
 
-  const handleModalOpen = () => {
-    setIsModalVisible(true)
+  // If removing permissions, we want to remove it from both the permissions and addedPermissions array and add it to the removedPermissions array if it doesn't exist there
+  const handleRemovedPermission = (
+    removedPermission: AuthAdminClientAllowedScope,
+  ) => {
+    const newPermissions = permissions.filter(
+      (item) => item.name !== removedPermission.name,
+    )
+    const newAddedScopes = addedScopes.filter(
+      (item) => item.name !== removedPermission.name,
+    )
+    setAddedScopes(newAddedScopes)
+    setPermissions(newPermissions)
+
+    setRemovedScopes((prevState) => {
+      if (
+        [...addedScopes, ...prevState].find(
+          (item) => item.name === removedPermission.name,
+        )
+      ) {
+        return prevState
```

Comment:
I don't understand why concat `addedScopes`?

Reply:
if you are talking about the [...addedScopes, ...prevState] then I am checking if it already exists in the removedScopes, or if it exists in addedScopes. Then I dont want to add it to the removedScopes state, then it just enough to remove it from the addedScopes array.

Reply:
If it was previously in the addedScopes it means that it was added from AddPermission.tsx and has not been sent to the server, so when it is removed, we dont want it in the array for removedScopes that gets sent to the server.